packages/contracts/contracts/extensions/SemaphoreZk3.sol

```diff
@@ -0,0 +1,133 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+
+import "../interfaces/ISemaphoreZk3.sol";
+import "../base/SemaphoreGroups.sol";
+import "../interfaces/ISemaphoreVerifier.sol";
+
+/// @title SemaphoreZK3 circles contract.
+/// @dev The following code allows you to create circles, add members and allow them to double signal.
+contract SemaphoreZk3 is ISemaphoreZk3, SemaphoreGroups {
+    /// @dev Gets a tree depth and returns its verifier address.
+    ISemaphoreVerifier internal verifier;
+
+    /// @dev Gets a circle id and returns the circle data.
+    mapping(uint256 => Circle) internal circles;
+
+    /// @dev Gets a nullifier hash and returns true or false.
+    /// It is used to prevent double-voting.
+    mapping(uint256 => bool) internal nullifierHashes;
```

Comment:
It might be necessary to have a set of nullifier hashes for each circle/group, like here: https://github.com/semaphore-protocol/semaphore/blob/b45c4f3510559f9c2b499074d1788e73ae2e271f/packages/contracts/contracts/interfaces/ISemaphoreVoting.sol#L26.

Otherwise, if the same identity were used in two groups with the same external nullifier, after consuming the first nullifier hash with the first proof with one group, that same identity could not generate another proof with the other group, since the nullifier hash would result to be already used.